torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -596,7 +642,10 @@ void TensorPipeAgent::respond(std::shared_ptr<tensorpipe::Pipe>& pipe) {
         threadPool_.run([this,
                          pipe,
                          messageId,
-                         requestMessage{std::move(requestMessage)}]() mutable {
+                         requestMessage{std::move(requestMessage)},
+                         streams{std::move(streams)}]() mutable {
+          // create guards again as this function runs on a different thread
+          auto guards = streamsToGuards(streams);
```

Comment:
The streams we receive from `pipeRead` do, at the moment, only contain streams for the devices on which the input tensors lived. However the user function may place the result tensors on another different device. I therefore think we should get a stream from the pool for all devices and set them all as current.